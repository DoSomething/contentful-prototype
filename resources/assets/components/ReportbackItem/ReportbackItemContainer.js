import { connect } from 'react-redux';
import { get } from 'lodash';
import ReportbackItem from './ReportbackItem';
import { isAuthenticated } from '../../selectors/user';
import { toggleReactionOn, toggleReactionOff } from '../../actions';

/**
 * Provide state from the Redux store as props for this component.
 */
const mapStateToProps = (state, props) => {
  const reportback = state.reportbacks.entities[props.id];

  if (!reportback) {
    return { isFetching: true };
  }

  const reportbackItem =
    state.reportbacks.itemEntities[reportback.reportback_items[0]];

  return {
    isFetching: false,
    id: reportbackItem.id,
    url: reportbackItem.media.uri,
    quantity: reportback.quantity,
    noun: get(state.campaign.additionalContent, 'noun'),
    firstName: reportback.user.first_name,
    reaction: reportbackItem.reaction,
    caption: reportbackItem.caption,
    isAuthenticated: isAuthenticated(state),
    reportback,
  };
};

/**
 * Provide pre-bound functions that allow the component to dispatch
 * actions to the Redux store as props for this component.
 */
const actionCreators = {
  toggleReactionOn,
  toggleReactionOff,
};

// Export the container component.
export default connect(mapStateToProps, actionCreators)(ReportbackItem);
