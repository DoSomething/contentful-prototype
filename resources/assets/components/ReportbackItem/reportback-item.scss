@import '../../scss/next-toolbox.scss';

.reportback-item {
  font-family: $primary-font-family;

  h4 {
    font-size: $font-regular;
    font-weight: $weight-bold;
  }

  p {
    color: $med-gray;
  }

  > .figure__media {
    background-color: $med-gray;

    & > img {
      width: 480px; // HACK: Expand smaller images to fit in card.
    }
  }

  > .figure__body p {
    word-break: break-all;
  }
}

.padded {
  padding: $base-spacing / 2;
}
