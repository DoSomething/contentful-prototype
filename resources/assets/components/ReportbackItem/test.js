import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import ReportbackItem from './ReportbackItem';

test('Reportback Item basic snapshot test', () => {
  const component = shallow(<ReportbackItem />);
  const tree = shallowToJson(component);
  expect(tree).toMatchSnapshot();
});

test('Reportback Item snapshot test', () => {
  const component = shallow(
    <ReportbackItem
      id="09251952"
      caption="Some awesmome caption"
      firstName="Luke Skywalker"
      noun={{ singular: 'lightsaber', plural: 'lightsabers' }}
      quantity={20}
      url="https://static1.comicvine.com/uploads/original/11118/111184078/5124660-6531264806-22292.jpg"
      reaction={{ id: '456', reacted: true, termId: '789', total: 30 }}
    />,
  );
  const tree = shallowToJson(component);
  expect(tree).toMatchSnapshot();
});
