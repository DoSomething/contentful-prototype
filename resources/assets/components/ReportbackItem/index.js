import PropTypes from 'prop-types';
import React from 'react';
import { Figure, BaseFigure, ReportbackFigure } from '../Figure';
import Reaction from '../Reaction';
import { pluralize } from '../../helpers';
import './reportback-item.scss';

// TODO: Use this for the puck connection.
//
// function getMetadataFromProps(props) {
//   return {
//     itemId: props.id,
//     quantity: props.quantity,
//     totalReactions: props.reaction ? props.reaction.total : null,
//     reportbackId: props.reportback ? props.reportback.id : null,
//     reportbackUser: props.reportback ? props.reportback.user : null,
//   };
// }

const ReportbackItem = (props) => {
  const { id, url, quantity, noun, caption, firstName, reaction = null, isFetching = false,
    basicDisplay = false, toggleReactionOn, toggleReactionOff } = props;

  const reactionElement = reaction ? (
    <Reaction
      active={reaction.reacted}
      total={reaction.total}
      onToggleOn={() => toggleReactionOn(id, reaction.termId)}
      onToggleOff={() => toggleReactionOff(id, reaction.id)}
    />
  ) : null;

  if (isFetching) {
    return (
      <Figure className="reportback-item" image="" alt="Loading...">
        <BaseFigure media={reactionElement} alignment="right" className="padded">
          <h4>Loading…</h4>
          <p className="footnote">…</p>
        </BaseFigure>
      </Figure>
    );
  }

  const meta = caption ? (<p className="italic">{caption}</p>) : null;

  return (
    <Figure image={url} alt={`${firstName}'s photo`}>
      <ReportbackFigure media={reactionElement} meta={meta} alignment="right" className="padded">
        {! basicDisplay && firstName ? <h4>{firstName}</h4> : null }
        {! basicDisplay && quantity ? <p className="footnote italic">{quantity} {pluralize(quantity, noun.singular, noun.plural)}</p> : null }
      </ReportbackFigure>
    </Figure>
  );
};

ReportbackItem.propTypes = {
  id: PropTypes.string,
  basicDisplay: PropTypes.bool,
  caption: PropTypes.string,
  firstName: PropTypes.string,
  isFetching: PropTypes.bool,
  noun: PropTypes.shape({
    singular: PropTypes.string,
    plural: PropTypes.string,
  }),
  quantity: PropTypes.number,
  reaction: PropTypes.shape({
    id: PropTypes.string,
    reacted: PropTypes.bool,
    termId: PropTypes.string,
    total: PropTypes.number,
  }),
  toggleReactionOff: PropTypes.func,
  toggleReactionOn: PropTypes.func,
  url: PropTypes.string,
};

// @TODO: The following props needed to be defined as undefined likely due to how we are
// implementing this component as a lazy loaded item. The props could not be set as required which
// thus meant we needed default props. However, setting the default props as null would cause the
// app to crash. Something to look into...
ReportbackItem.defaultProps = {
  id: undefined,
  basicDisplay: false,
  caption: undefined,
  firstName: 'A Doer',
  isFetching: false,
  noun: {
    singular: 'item',
    plural: 'items',
  },
  quantity: undefined,
  reaction: null,
  url: undefined,
  toggleReactionOff: () => {},
  toggleReactionOn: () => {},
};

export default ReportbackItem;
