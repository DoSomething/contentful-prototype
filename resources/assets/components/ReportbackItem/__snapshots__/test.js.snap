// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reportback Item basic snapshot test 1`] = `
<div
  className="reportback-item"
>
  <Figure
    alt="A Doer's photo"
    image={null}
    imageClassName={null}
  >
    <BaseFigure
      alignment="right"
      className="padded"
      media={null}
      size={null}
      verticalAlignment={null}
    >
      <h4>
        A Doer
      </h4>
    </BaseFigure>
  </Figure>
  <div
    className="caption"
  />
</div>
`;

exports[`Reportback Item snapshot test 1`] = `
<div
  className="reportback-item"
>
  <Figure
    alt="Luke Skywalker's photo"
    image="https://static1.comicvine.com/uploads/original/11118/111184078/5124660-6531264806-22292.jpg"
    imageClassName={null}
  >
    <BaseFigure
      alignment="right"
      className="padded"
      media={
        <Reaction
          active={true}
          onToggleOff={[Function]}
          onToggleOn={[Function]}
          total={30}
        />
      }
      size={null}
      verticalAlignment={null}
    >
      <h4>
        Luke Skywalker
      </h4>
      <p
        className="footnote italic"
      >
        20
         
        lightsabers
      </p>
    </BaseFigure>
  </Figure>
  <div
    className="caption"
  >
    <p
      className="italic"
    >
      Some awesmome caption
    </p>
  </div>
</div>
`;
